base/essentials.jl

```diff
@@ -257,7 +257,7 @@ end
 
 function rewrap_unionall(t::Core.TypeofVararg, @nospecialize(u))
     isdefined(t, :T) || return t
-    if !isdefined(t, :N) || t.N === u.var
+    if !isdefined(t, :N) || (isa(u, UnionAll) && t.N === u.var)
         return Vararg{rewrap_unionall(t.T, u)}
     end
     Vararg{rewrap_unionall(t.T, u), t.N}
```

Comment:
nit: make this use the same code pattern as the original method it replaced, and DRY slightly:
```suggestion
    if !isa(u, UnionAll)
        return t
    end
    T = rewrap_unionall(t.T, u)
    if !isdefined(t, :N) || t.N === u.var
        return Vararg{T}
    end
    return Vararg{T, t.N}
```